File: ios/Sources/ForgeIOSRuntime/NativeDateRangePreset.swift
import Foundation

public enum NativeDateRangePreset {
    public static func patch(item: ItemDef, value: JSONValue, metrics: [String: JSONValue] = [:], now: Date = Date()) -> [String: JSONValue]? {
        func setting(_ key: String, fallback: String) -> String {
            if let selector = item.properties[key + "Selector"]?.stringValue {
                let resolved = SelectorUtil.resolve(metrics, selector: selector)
                if let text = resolved as? String { return text }
                if let json = resolved as? JSONValue, let text = json.stringValue { return text }
            }
            return item.properties[key]?.stringValue ?? fallback
        }
        guard let range = resolve(NativeWidgetContract.text(value), now: now, timeZone: setting("timeZone", fallback: "UTC"), lifetimeStart: setting("lifetimeStart", fallback: "2026-01-01")) else { return nil }
        return [item.properties["startField"]?.stringValue ?? "customDateStart": range["start"]!, item.properties["endField"]?.stringValue ?? "customDateEnd": range["end"]!, item.properties["granularityField"]?.stringValue ?? "granularity": range["granularity"]!]
    }
    public static func resolve(_ value: String, now: Date = Date(), timeZone: String = "UTC", lifetimeStart: String = "2026-01-01") -> [String: JSONValue]? {
        var calendar = Calendar(identifier: .gregorian); calendar.timeZone = TimeZone(identifier: timeZone) ?? TimeZone(secondsFromGMT: 0)!
        let today = calendar.startOfDay(for: now)
        let components = calendar.dateComponents([.year, .month, .day], from: today)
        func date(_ year: Int, _ month: Int, _ day: Int) -> Date { calendar.date(from: DateComponents(year: year, month: month, day: day))! }
        let year = components.year!, month = components.month!, day = components.day!
        var start = today, end = today
        switch value.lowercased() {
        case "today": break
        case "yesterday": start = calendar.date(byAdding: .day, value: -1, to: today)!; end = start
        case "week": start = calendar.date(byAdding: .day, value: -6, to: today)!
        case "month": start = calendar.date(byAdding: .month, value: -1, to: today)!
        case "month_to_date": start = date(year,month,1)
        case "last_month": start = date(year,month-1,1); end = calendar.date(byAdding: .day, value: -1, to: date(year,month,1))!
        case "quarter_to_date": start = date(year,((month-1)/3)*3+1,1)
        case "year_to_date": start = date(year,1,1)
        case "year": start = date(year-1,month,day)
        case "last_year": start = date(year-1,1,1); end = date(year-1,12,31)
        case "lifetime": break
        default: return nil
        }
        let formatter = DateFormatter(); formatter.locale = Locale(identifier: "en_US_POSIX"); formatter.timeZone = calendar.timeZone; formatter.dateFormat = "yyyy-MM-dd"
        let days = (calendar.dateComponents([.day], from: start, to: end).day ?? 0) + 1
        return ["start": .string(value.lowercased() == "lifetime" ? lifetimeStart : formatter.string(from: start)), "end": .string(formatter.string(from: end)), "granularity": .string(value.lowercased() != "lifetime" && days <= 2 ? "hour" : "day")]
    }
}

File: android/sdk/src/main/java/com/viant/forgeandroid/runtime/NativeDateRangePreset.kt
package com.viant.forgeandroid.runtime

import java.time.*
import java.time.temporal.ChronoUnit

object NativeDateRangePreset {
    fun patch(item: ItemDef, value: String, metrics: Map<String, Any?> = emptyMap(), now: Instant = Instant.now()): Map<String, Any?>? {
        fun property(key: String) = (item.properties[key] as? kotlinx.serialization.json.JsonPrimitive)?.content
        val zone = (item.timeZoneSelector ?: property("timeZoneSelector"))?.let { SelectorUtil.resolve(metrics,it)?.toString() } ?: item.timeZone ?: property("timeZone") ?: "UTC"
        val lifetime = (item.lifetimeStartSelector ?: property("lifetimeStartSelector"))?.let { SelectorUtil.resolve(metrics,it)?.toString() } ?: item.lifetimeStart ?: property("lifetimeStart") ?: "2026-01-01"
        val range = resolve(value, now, zone, lifetime) ?: return null
        return mapOf((item.startField ?: property("startField") ?: "customDateStart") to range["start"], (item.endField ?: property("endField") ?: "customDateEnd") to range["end"], (item.granularityField ?: property("granularityField") ?: "granularity") to range["granularity"])
    }
    fun resolve(value: String, now: Instant = Instant.now(), timeZone: String = "UTC", lifetimeStart: String = "2026-01-01"): Map<String, Any?>? {
        val zone = runCatching { ZoneId.of(timeZone) }.getOrDefault(ZoneOffset.UTC)
        val today = now.atZone(zone).toLocalDate()
        var end = today
        val start = when (value.lowercase()) {
            "today" -> today
            "yesterday" -> today.minusDays(1).also { end = it }
            "week" -> today.minusDays(6)
            "month" -> today.minusMonths(1)
            "month_to_date" -> today.withDayOfMonth(1)
            "last_month" -> today.withDayOfMonth(1).minusMonths(1).also { end = today.withDayOfMonth(1).minusDays(1) }
            "quarter_to_date" -> today.withMonth(((today.monthValue-1)/3)*3+1).withDayOfMonth(1)
            "year_to_date" -> today.withDayOfYear(1)
            "year" -> LocalDate.of(today.year-1,today.monthValue,1).plusDays((today.dayOfMonth-1).toLong())
            "last_year" -> LocalDate.of(today.year-1,1,1).also { end = LocalDate.of(today.year-1,12,31) }
            "lifetime" -> today
            else -> return null
        }
        return mapOf("start" to if (value.lowercase() == "lifetime") lifetimeStart else start.toString(), "end" to end.toString(), "granularity" to if (value.lowercase() != "lifetime" && ChronoUnit.DAYS.between(start,end)+1 <= 2) "hour" else "day")
    }
}
